src/cached_object_store.rs

```diff
@@ -762,6 +798,214 @@ impl LocalCacheEntry for FsCacheEntry {
     }
 }
 
+#[derive(Debug)]
+struct FsCacheEvictor {
+    root_folder: std::path::PathBuf,
+    limit_bytes: u64,
+    tx: tokio::sync::mpsc::Sender<u64>,
+    rx: Mutex<Option<tokio::sync::mpsc::Receiver<u64>>>,
+}
+
+impl FsCacheEvictor {
+    pub fn new(root_folder: std::path::PathBuf, limit_bytes: u64) -> Self {
+        let (tx, rx) = tokio::sync::mpsc::channel(100);
+        Self {
+            root_folder,
+            limit_bytes,
+            tx,
+            rx: Mutex::new(Some(rx)),
+        }
+    }
+
+    async fn start(&self) {
+        // if the limit_bytes is 0, do nothing
+        if self.limit_bytes == 0 {
+            return;
+        }
+
+        let rx = self
+            .rx
+            .lock()
+            .await
+            .take()
+            .expect("evictor already started");
+        tokio::spawn(Self::background(
+            self.root_folder.clone(),
+            self.limit_bytes,
+            rx,
+        ))
+        .await
+        .ok();
+    }
+
+    async fn background(
+        root_folder: std::path::PathBuf,
+        limit_bytes: u64,
+        mut rx: tokio::sync::mpsc::Receiver<u64>,
+    ) {
+        let start_time = std::time::Instant::now();
+        let scanned_bytes = match FsCacheEvictorInner::scan_cache_folder_bytes(&root_folder).await {
+            Ok(scanned_bytes) => scanned_bytes,
+            Err(err) => {
+                warn!(
+                    "evictor: unexpected error on scaning the cache folder after {}s, stop the evictor: {}",
+                    start_time.elapsed().as_secs(),
+                    err
+                );
+                return;
+            }
+        };
+        let inner = FsCacheEvictorInner {
+            root_folder,
+            tracked_bytes: AtomicU64::new(scanned_bytes as u64),
+            limit_bytes,
+            evict_buffer: Mutex::new(VecDeque::new()),
+        };
+        while let Some(bytes) = rx.recv().await {
+            inner.maybe_evict(bytes).await;
+        }
+    }
+
+    pub async fn maybe_evict(&self, bytes: u64) {
+        if self.limit_bytes == 0 {
+            return;
+        }
+
+        self.tx.send(bytes).await.ok();
+    }
+}
+
+struct FsCacheEvictorInner {
+    root_folder: std::path::PathBuf,
+    tracked_bytes: AtomicU64,
+    limit_bytes: u64,
+    evict_buffer: Mutex<VecDeque<std::path::PathBuf>>,
+}
+
+impl FsCacheEvictorInner {
+    pub async fn maybe_evict(&self, bytes: u64) {
+        self.tracked_bytes.fetch_add(bytes, Ordering::SeqCst);
+        if self.tracked_bytes.load(Ordering::Relaxed) <= self.limit_bytes {
+            return;
+        }
+
+        let evict_factor = 10;
+        for _ in 0..evict_factor {
```

Comment:
yes, i hope this could reduce the frequency of executing evictions, it might not bad to allow executing evictions in a batch manner.

（i'm not certain if there's an existing term to describe this behaviour..